src/tsdb/prometheus.act

```diff
@@ -0,0 +1,147 @@
+                json_dict = { "name": "default" }
+                #http_client.post("/api/v1/services/prometheusdb/namespace/ready", json.encode(json_dict).encode(), lambda c, r: _on_ready_response(c, r, timeout))
+
+            def _on_ready_response(client: http.Client, response: http.Response, timeout: int):
+                log.trace("Namespace ready request:" + response.body.decode(), None)
+
+                if response.status != 200 and timeout > 0:
+                    after 1: _post_ready(timeout - 1)
+                else:
+                    await async on_response(self, response.status == 200)
+
+            _post_ready(20)
+
+        json_dict = { "type": "local", "namespaceName": "default", "retentionTime": "12h" }
+        #http_client.post("/api/v1/database/create", json.encode(json_dict).encode(), _on_response)
+
+    def send_metric(metric: Metric, on_response: action(Client, bool) -> None):
+
+        #mtags = {}
+        #for k, v in metric.tags.items():
+        #    mtags[k] = v
+
+        #mtags["__name__"] = metric.name
+        #json_dict = { "tags": mtags, "value": metric.value }
+
+        #json_dict["timestamp"] = str(metric.timestamp)
+
+        #def _on_response(client: http.Client, response: http.Response):
+        #    log.debug("HTTP response status: " + str(response.status), None)
+        #    log.trace(response.body.decode(), None)
+        #    await async on_response(self, response.status == 200)
+
+        #json_string = json.encode(json_dict)
+        #log.trace(json_string, None)
+
+        #if json_string is not None:
+        #    http_client.post("/api/v1/json/write", json_string.encode(), _on_response)
+        send_metrics([metric], on_response)
+
+    def send_metrics(metrics: list[Metric], on_response: action(Client, bool) -> None):
+        def _on_response(client: http.Client, response: http.Response):
+            log.debug("HTTP response status: " + str(response.status), None)
+            log.debug(response.body.decode(), None)
+
+        metric_data: list[(list[(str, str)], list[(float, int)])] = []
+        metric_dict: dict[str, list[Metric]] = {}
+        
+        for metric in metrics:
+            series_id = metric.name + str(metric.tags)
+
+            if series_id not in list(metric_dict.keys()):
+                 metric_dict[series_id] = []
+
+            metric_dict[series_id].append(metric)
+
+        for _, metric_series in metric_dict.items():
+            labels = [("__name__", metric_series[0].name)]
+            for name, value in metric_series[0].tags.items():
+                labels.append((name, value))
+
+            samples = []
+            for metric in metric_series:
+                samples.append((metric.value, metric.timestamp * 1000))
+
+            metric_data.append((labels, samples))
+
+        log.debug("metric_data: " + str(metric_data), None)
+        metric_proto = pack(metric_data)
+        log.debug("metric_proto: " + str(metric_proto), None)
```

Comment:
Same as above, do not str format stuff into the log message itself. Place data in the second arg.